devito/passes/iet/mpi.py

```diff
@@ -59,120 +59,116 @@ def _hoist_halospots(iet):
 
-    def rule1(dep, hs, loc_indices):
-        # TODO This is apparently never hit, but feeling uncomfortable to remove it
-        return (dep.is_regular and
-                dep.read is not None and
-                all(not any(dep.read.touched_halo(d.root)) for d in dep.cause))
+        for iter in iters:
+            # If also merge-able we can start hoisting the latter
+            for dep in scopes[iter].d_flow.project(f):
+                if not any(r(dep, hs1, v.loc_indices) for r in merge_rules()):
+                    break
+            else:
+                hse = hs1.halo_scheme.fmapper[f]
+                raw_loc_indices = {}
+                # Entering here means we can lift, and we need to update
+                # the loc_indices with known values
+                # TODO: Can I get this in a more elegant way?
+                for d in hse.loc_indices:
+                    if hse.loc_indices[d].is_Symbol:
+                        assert d in hse.loc_indices[d]._defines
+                        root_min = hse.loc_indices[d].symbolic_min
+                        new_min = root_min.subs(hse.loc_indices[d].root,
+                                                hse.loc_indices[d].root.symbolic_min)
+                        raw_loc_indices[d] = new_min
+                    else:
+                        assert d.symbolic_min in hse.loc_indices[d].free_symbols
+                        raw_loc_indices[d] = hse.loc_indices[d]
 
-    def rule2(dep, hs, loc_indices):
-        # E.g., `dep=W<f,[t1, x+1]> -> R<f,[t1, xl+1]>` and `loc_indices={t: t0}` => True
-        return any(dep.distance_mapper[d] == 0 and dep.source[d] is not v
-                   for d, v in loc_indices.items())
+                hse = hse.rebuild(loc_indices=frozendict(raw_loc_indices))
+                hs1.halo_scheme.fmapper[f] = hse
 
-    rules = [rule0, rule1, rule2]
+                hsmapper[hs1] = hsmapper.get(hs1, hs1.halo_scheme).drop(f)
+                imapper[iter].append(hs1.halo_scheme.project(f))
 
-    # Analysis
-    cond_mapper = MapHaloSpots().visit(iet)
-    cond_mapper = {hs: {i for i in v if i.is_Conditional and
-                        not isinstance(i.condition, GuardFactorEq)}
-                   for hs, v in cond_mapper.items()}
 
-    iter_mapper = MapNodes(Iteration, HaloSpot, 'immediate').visit(iet)
+def _merge_halospots(iet):
+    """
+    Merge HaloSpots on the same Iteration tree level where all data dependencies
+    would be honored.
+    """
+
+    # Analysis
+    cond_mapper = _create_cond_mapper(iet)
 
     mapper = {}
-    for i, halo_spots in iter_mapper.items():
-        if i is None or len(halo_spots) <= 1:
+    for iter, halo_spots in MapNodes(Iteration, HaloSpot, 'immediate').visit(iet).items():
+        if iter is None or len(halo_spots) <= 1:
             continue
 
-        scope = Scope([e.expr for e in FindNodes(Expression).visit(i)])
+        scope = Scope([e.expr for e in FindNodes(Expression).visit(iter)])
```

Comment:
don't use the word "iter" since it has a special meaning in python

Reply:
fixed everywhere